Local build automatizated scripts

charliemartinez opened the pull request.

manongjohn commented on ci-scripts/linux/local-build/linux_build.sh:
```diff
@@ -0,0 +1,411 @@
+
+# Compile:
+
+_cloningTahoma # Cloning the GIT Tree
```
Seems kind of strange to do this step, considering you will get this script when you clone the repository so it should already be there. 🙂

charliemartinez replied:
Hehe, well, the script stays true to the documentation...

![imagen](https://github.com/user-attachments/assets/f8aa41f9-7cbf-4d58-a431-49875b0b8748)

I think maybe I'm not the only one who relies heavily on GitHub's graphical interface and downloads files with the mouse. In any case, the interesting thing about this function is that no matter from which path you run the script, it will download the code right there and compile it.

charliemartinez replied:
Corrected

manongjohn commented on ci-scripts/linux/local-build/README.md:
```diff
@@ -0,0 +1,35 @@
+# Linux Local Build Scripts
```
It might be worthwhile to make a suggestion to use these scripts and link to this file in [tahoma2d](https://github.com/tahoma2d/tahoma2d/tree/master)/[doc](https://github.com/tahoma2d/tahoma2d/tree/master/doc)
/how_to_build_linux.md

charliemartinez replied:
Hehe, well, the script stays true to the documentation...

![imagen](https://github.com/user-attachments/assets/f8aa41f9-7cbf-4d58-a431-49875b0b8748)

I think maybe I'm not the only one who relies heavily on GitHub's graphical interface and downloads files with the mouse. In any case, the interesting thing about this function is that no matter from which path you run the script, it will download the code right there and compile it.

manongjohn replied:
That would be bad if you run the script within the repo.  See my other comment.

charliemartinez replied:
> That would be bad if you run the script within the repo. See my other comment.

You’re right.

manongjohn commented on ci-scripts/linux/local-build/README.md:
````diff
@@ -0,0 +1,35 @@
+1) Run the `linux_depends.sh` script with root permissions to install the necessary dependencies:
+
+   ```bash
+   chmod +x ./linux_depends.sh
````
Considering these scripts are deep in the repo folder structure and typically the build instructions are done from the root of the repo, the instructions should probably say to be in the root of the tahoma2d repo and provide path from there to the scripts

```
chmod +x ./ci-scripts/linux/local-build/linux_depends.sh
sudo ./ci-scripts/linux/local-build/linux_depends.sh
```

charliemartinez replied:
> Considering these scripts are deep in the repo folder structure and typically the build instructions are done from the root of the repo, the instructions should probably say to be in the root of the tahoma2d repo and provide path from there to the scripts

These scripts automate the repository download, meaning you only need to download the scripts and place them anywhere. The linux_build.sh script will handle cloning the repository in the location where it is executed and will store it in a tahoma2d folder, regardless of the root directory (with function _clone()).

manongjohn replied:
If someone is going to be building T2D from source, they are likely going to clone the repo anyway, especially if they are following the posted build instructions.

If these scripts were available outside of the repo, it makes sense do what you are doing but with it already being in the repo, it doesn't.   Can just as easily clone the repo, as per instructions, and then run the script that is already in the repo to do the build/rebuild. 

The script shouldn't be recloning the repo.  This would be especially bad if they run the script inside the repo.  They would have a repo within the repo.  Git would consider the entire contents of the inner repo as new stuff to be added to the repo.

Additionally, it think it should only ever clone the repo once, and skip it if already there. I would find it annoying to have to keep answering 'no' every time I wanted to rebuild using your scripts.  Updating the existing repo can be easily done with git command and if really needed, they can manually delete the entire folder and start from scratch.

charliemartinez replied:
You’re right. I just changed the code to offer a more practical solution: the script will now look for the tahoma2d folder at the same level. If it finds it, it will compile. If it doesn’t find it, it will explain that it has to be at the same level as the script, give the option to search for it again, and then offer the option to automatically clone the repository.

charliemartinez replied:
And in my last commit, I added the option for the user to provide the path directly. This menu will only appear when it's truly necessary (if the program finds the "tahoma2d" folder, it will simply proceed with the compilation). Now I'm done—unless you find any other mistake on my part, which I’ll gladly fix. Best regards.

manongjohn commented on ci-scripts/linux/local-build/linux_build.sh:
```diff
@@ -0,0 +1,411 @@
+    fi
+    
+    if [ -d "$BUILD_DIR" ]; then # If it exists, it deletes it to create a new build. *
+    rm -rf "$BUILD_DIR" 
```
I dont necessarily recommend this. A full build could take a bit of time to compile.  If they make a minor change, if it reused the current build directory, the build may only take a few minutes.  I would be very annoyed if every change I made caused a full rebuild umnecessarily.

charliemartinez replied:
> I dont necessarily recommend this. A full build could take a bit of time to compile. If they make a minor change, if it reused the current build directory, the build may only take a few minutes. I would be very annoyed if every change I made caused a full rebuild umnecessarily.

I agree, it has been tedious for me to recompile everything every time I modified something until I found what I wanted to do today... Maybe I haven't fully understood how it works... Are you saying that I shouldn't make changes in the src folder, but instead in tahoma2d/toonz/build/toonz/? And then I could do a partial build as long as I don't overwrite the build folder?

manongjohn replied:
No. You still have to make changes in the source folders, however, it is unnecessary to delete the build directory and start the build from scratch every time. 

Compilers know what's changed from the prior compilation and will only recompile the parts that were impacted by the change, thereby making rebuilding faster.  Every time you delete the build directory, you remove all of the information it needs to do a quick recompile of only what's changed.   If I corrected a typo and had to wait 1 hour for it to recompile, I'd be extremely frustrated and probably wouldn't code anymore.

The only time you should really be deleting the build directory is if there is a change to major dependencies, like say upgrading Qt version or boost since the paths to the newer libraries is hard coded into the build environment when it was 1st created.

charliemartinez replied:
Great! I deleted those lines. I've learned a lot today, thanks!!!

limitedAtonement commented on doc/how_to_build_linux.md:
```diff
@@ -1,5 +1,9 @@
+
 ## Required software
 
 Building Tahoma2D from source requires the following dependencies:
```
This document suggests you'll need to get `blas` and `cblas` from AUR, but the new scripts added in this pull request do not do that. Can we make the documentation match the scripts?

limitedAtonement commented on ci-scripts/linux/local-build/linux_build.sh:
```diff
@@ -0,0 +1,559 @@
+function _checkdepends() {
+	# Check if dependencies have been installed
+	if [ ! -e "$SCRIPT_DIR/ok-packages" ]; then
+	echo""
```
Some `if` blocks are indented and some are not. They should probably all be indented for readability.

limitedAtonement commented on ci-scripts/linux/local-build/linux_depends.sh:
```diff
@@ -0,0 +1,268 @@
+#!/bin/bash
```
This file should be executable.

limitedAtonement commented on ci-scripts/linux/local-build/linux_build.sh:
```diff
@@ -0,0 +1,559 @@
+#!/bin/bash
```
This file should be executable.

limitedAtonement commented on ci-scripts/linux/local-build/linux_depends.sh:
```diff
@@ -0,0 +1,268 @@
+_rootcheck      # Check for root privileges
+_verifpackages  # Verify the existence of the test file
+_warning		# Show a warning about process
+_update         # Update the system
```
Why would you update the user's system? This seems well out of scope for an installation script.

limitedAtonement commented on ci-scripts/linux/local-build/linux_depends.sh:
```diff
@@ -0,0 +1,268 @@
+
+function _depends() {
+	# Detect and install dependencies based on the system's package manager
+	clear	
```
Clearing the terminal is not user friendly.

limitedAtonement commented on ci-scripts/linux/local-build/linux_depends.sh:
```diff
@@ -0,0 +1,268 @@
+	echo -e "\n\n"
+    for paquetes in dialog base-devel git cmake boost boost-libs qt5-base qt5-svg qt5-script \
+        qt5-tools qt5-multimedia lz4 lzo libjpeg-turbo glew freeglut freetype2 blas cblas \
+        superlu libmypaint; do pacman -S --needed
```
`pacman -S --needed` doesn't do anything. (You need to specify a package!) Probably better would be:

```
pacman -S --needed dialog base-devel git ... libmypaint; || exit;
```

limitedAtonement commented on ci-scripts/linux/local-build/linux_depends.sh:
```diff
@@ -0,0 +1,268 @@
+        qt5-tools qt5-multimedia lz4 lzo libjpeg-turbo glew freeglut freetype2 blas cblas \
+        superlu libmypaint; do pacman -S --needed
+    done
+    touch /tmp/ok-packages
```
This script touches `/tmp/ok-packages`, but `linux_build.sh` checks `"$SCRIPT_DIR/ok-packages"`. That's not going to work.

limitedAtonement commented on ci-scripts/linux/local-build/linux_depends.sh:
```diff
@@ -0,0 +1,268 @@
+    local clave=$1 # Store the key for the requested message
+    declare -A msg=( # Translation messages for different languages
+    
+    	[es_INTERNET]="Este programar equiere conexión a internet"
```
This line uses spaces and tabs for indentation, but most lines just use spaces.

limitedAtonement commented on ci-scripts/linux/local-build/linux_build.sh:
```diff
@@ -0,0 +1,559 @@
+    local TAHOMA_DIR="./tahoma2d"
+    
+    # Check if the folder exists in the default location
+    if [ -d "$TAHOMA_DIR" ]; then
```
When I run this script like this:

1. `git clone https://github.com/tahoma2d/tahoma2d`
2. `cd tahoma2d/ci-scripts/linux/local-build`
3. `chmod +x ./linux_*`
4. `./linux_depends.sh`
5. `./linux_build.sh`

I get an error about not being able to find Tahoma. This seems like it should be a supported use-case.

limitedAtonement commented on ci-scripts/linux/local-build/linux_build.sh:
```diff
@@ -0,0 +1,559 @@
+        echo ''
+        echo "$(_msg ROOTCHECK)"
+        echo ''
+        exit 1
```
Warning that the user is root seems reasonable, but completely aborting the installation seems heavy-handed. For instance, I'm building in docker as root and I understand the consequences.

limitedAtonement commented on ci-scripts/linux/local-build/linux_build.sh:
```diff
@@ -0,0 +1,559 @@
+        echo "  1) $(_msg SEARCH_AGAIN)"
+        echo "  2) $(_msg AUTOMATIC_CLONING)"
+        echo "  3) $(_msg SPECIFY_DIRECTORY)"
+        echo "  4) $(_msg EXIT)"
```
Elsewhere, you use `dialog` to present options to the user. Why not here?

limitedAtonement commented on ci-scripts/linux/local-build/linux_build.sh:
```diff
@@ -0,0 +1,559 @@
+    if command -v dnf >/dev/null 2>&1 || command -v yum >/dev/null 2>&1; then
+        cmake "$SOURCES_DIR" -DCMAKE_INSTALL_PREFIX="$BUILD_DIR" -DSUPERLU_INCLUDE_DIR=/usr/include/SuperLU
+    else
+        cmake "$SOURCES_DIR" -DCMAKE_INSTALL_PREFIX="$BUILD_DIR"
```
This command failed for me, and it took way too long to figure out what went wrong because of all the `clear` and `dialog` that hid the error. You need to check the return code of most commands that are executed to see if they failed and decide whether you need to abandon the script.

limitedAtonement commented on ci-scripts/linux/local-build/README.md:
```diff
@@ -0,0 +1,39 @@
+# Linux Local Build Scripts
```
(This comment has nothing to do with this line of code. I'm just make a "conversation" that can be "resolved".)

By the way, the arch linux build is more or less doomed to fail because the code doesn't compile. To reproduce, run the following script:

```
cd tahoma2d;
docker run --rm -ti -v "$PWD":/src archlinux bash -i;
```
where `tahoma2d` is the cloned git repository. Once inside that docker container, run the following:

```
pacman -Syu base-devel git cmake boost boost-libs qt5-base qt5-svg qt5-script qt5-tools qt5-multimedia lz4 lzo libjpeg-turbo glew freeglut freetype2 blas cblas superlu libmypaint pipewire-jack qt5-serialport opencv;
mkdir build;
cd build;
cmake ../src/toonz/sources;
make;
```
For me it fails with
```
/src/toonz/sources/include/tthreadmessage.h: In constructor ‘TThread::Mutex::Mutex()’:
/src/toonz/sources/include/tthreadmessage.h:87:28: error: ‘Recursive’ is not a member of ‘QMutex’
   87 |   Mutex() : QMutex(QMutex::Recursive) {}
      |                            ^~~~~~~~~
```

(It should be `QRecursiveMutex()`.) This happens in many places across the code base and there are other compilation errors, too.

charliemartinez commented:
I will present a more polished version. However, what concerns me the most is that despite this script following the documented build steps, I’ve discovered something that doesn’t work with the webcam (and I haven’t tried it with the DSLR yet).

https://github.com/tahoma2d/tahoma2d/issues/1751

limitedAtonement commented:
@charliemartinez Would you create a ticket to organize this effort? It's best to design a feature and decide how everything should be structured in an Issue. I think it would be best to start with fixing the README build instructions. Once those are solid and work, automating it in a script should be very straightforward and nearly write itself.

charliemartinez commented:
> Would you create a ticket to organize this effort? It's best to design a feature and decide how everything should be structured in an Issue. I think it would be best to start with fixing the README build instructions. Once those are solid and work, automating it in a script should be very straightforward and nearly write itself.

Hello, thanks, I don't have much time and I dedicated all of yesterday and today to this script, because I really need it to quickly test things when maintaining the Tahoma2D packages for the Quirinux distro. Quirinux is based on Devuan, so I can test everything related to installation on Debian. The dependency issues and parameters that were failing on Debian seem to be fixed, and I already have a new version of the script ready to upload (I'm fixing the translations and, as much as possible, the indentation, so it won't hurt your eyes). I don't think it's necessary to create a roadmap or over-plan these scripts. The changes required to make them work better on Arch, Fedora, and OpenSuse won't be complicated to incorporate, although users working with those distros will be able to update the README.md better than I can. I want to clarify that these are more interactive programs than scripts, and in this version I'm about to upload, I kept using dialog and unified the criteria in linux_build.sh, although to compensate, I removed all the clear commands.

P.S.
I incorporated several improvements that I will publish soon, perhaps the most interesting one: incorporating the repository path as a variable and a menu that displays the available branches to choose which one to compile. I also separated the logic of the scripts from the text by creating a variables file with all the translations for each script, and another file for the path variables.

charliemartinez commented:
@manongjohn @limitedAtonement 

I did the best I could, and it compiles. I added parameters to compile libraries because, at least on Debian, installing dependencies seems to be useless. Even so, with --fmpeg and --opencv the webcam works, although with --gphoto gphoto compiles, my Canon camera does not appear in the menu.

I think it's quite useful for quick testing and modifying some parts of the code, but it's far from being useful as a preliminary step to creating a .deb that takes advantage of the system libraries. I'm afraid that, for now, the .deb packages I generate for Quirinux will continue to be extractions from the .AppImage.

https://github.com/tahoma2d/tahoma2d/pull/1755
